### PR TITLE
Add default of threshold to fix #614

### LINKED ISSUE #614: Error deserializing Scikit-learn LogisticRegression with Spark
I am training and serializing a scikit-learn Pipeline that contains a FeatureExtractor and a LogisticRegression, and I am using Spark to deserialize it in Python. But when I tried to run predict function on the deserialized pipeline model, I encountered an error saying `java.util.NoSuchElementException: Failed to find a default value for threshold`. Below is the full error log:

```
answer = 'xro255', gateway_client = <py4j.java_gateway.GatewayClient object at 0x7efc589eb0f0>, target_id = 'o248', name = 'collectToPython'

    def get_return_value(answer, gateway_client, target_id=None, name=None):
        """Converts an answer received from the Java gateway into a Python object.

        For example, string representation of integers are converted to Python
        integer, string representation of objects are converted to JavaObject
        instances, etc.

        :param answer: the string returned by the Java gateway
        :param gateway_client: the gateway client used to communicate with the Java
            Gateway. Only necessary if the answer is a reference (e.g., object,
            list, map)
        :param target_id: the name of the object from which the answer comes from
            (e.g., *object1* in `object1.hello()`). Optional.
        :param name: the name of the member from which the answer comes from
            (e.g., *hello* in `object1.hello()`). Optional.
        """
        if is_error(answer)[0]:
            if len(answer) > 1:
                type = answer[1]
                value = OUTPUT_CONVERTER[type](answer[2:], gateway_client)
                if answer[1] == REFERENCE_TYPE:
                    raise Py4JJavaError(
                        "An error occurred while calling {0}{1}{2}.\n".
>                       format(target_id, ".", name), value)
E                   py4j.protocol.Py4JJavaError: An error occurred while calling o248.collectToPython.
E                   : org.apache.spark.SparkException: Job aborted due to stage failure: Task 1 in stage 1.0 failed 1 times, most recent failure: Lost task 1.0 in stage 1.0 (TID 3, localhost, executor driver): org.apache.spark.SparkException: Failed to execute user defined function($anonfun$4: (struct<type:tinyint,size:int,indices:array<int>,values:array<double>>) => double)
E                   	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.processNext(Unknown Source)
E                   	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
E                   	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anonfun$13$$anon$1.hasNext(WholeStageCodegenExec.scala:636)
E                   	at org.apache.spark.sql.execution.SparkPlan$$anonfun$2.apply(SparkPlan.scala:255)
E                   	at org.apache.spark.sql.execution.SparkPlan$$anonfun$2.apply(SparkPlan.scala:247)
E                   	at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$24.apply(RDD.scala:836)
E                   	at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$24.apply(RDD.scala:836)
E                   	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
E                   	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)
E                   	at org.apache.spark.rdd.RDD.iterator(RDD.scala:288)
E                   	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
E                   	at org.apache.spark.scheduler.Task.run(Task.scala:121)
E                   	at org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)
E                   	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)
E                   	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)
E                   	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
E                   	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
E                   	at java.lang.Thread.run(Thread.java:748)
E                   Caused by: java.util.NoSuchElementException: Failed to find a default value for threshold
E                   	at org.apache.spark.ml.param.Params$$anonfun$getOrDefault$2.apply(params.scala:780)
E                   	at org.apache.spark.ml.param.Params$$anonfun$getOrDefault$2.apply(params.scala:780)
E                   	at scala.Option.getOrElse(Option.scala:121)
E                   	at org.apache.spark.ml.param.Params$class.getOrDefault(params.scala:779)
E                   	at org.apache.spark.ml.PipelineStage.getOrDefault(Pipeline.scala:42)
E                   	at org.apache.spark.ml.param.Params$class.$(params.scala:786)
E                   	at org.apache.spark.ml.PipelineStage.$(Pipeline.scala:42)
E                   	at org.apache.spark.ml.classification.LogisticRegressionParams$class.getThreshold(LogisticRegression.scala:122)
E                   	at org.apache.spark.ml.classification.LogisticRegressionModel.getThreshold(LogisticRegression.scala:991)
E                   	at org.apache.spark.ml.classification.LogisticRegressionModel.raw2prediction(LogisticRegression.scala:1174)
E                   	at org.apache.spark.ml.classification.ProbabilisticClassificationModel$$anonfun$4.apply(ProbabilisticClassifier.scala:136)
E                   	at org.apache.spark.ml.classification.ProbabilisticClassificationModel$$anonfun$4.apply(ProbabilisticClassifier.scala:136)
E                   	... 18 more
E
E                   Driver stacktrace:
E                   	at org.apache.spark.scheduler.DAGScheduler.org$apache$spark$scheduler$DAGScheduler$$failJobAndIndependentStages(DAGScheduler.scala:1889)
E                   	at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1877)
E                   	at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1876)
E                   	at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
E                   	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
E                   	at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:1876)
E                   	at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:926)
E                   	at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:926)
E                   	at scala.Option.foreach(Option.scala:257)
E                   	at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:926)
E                   	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:2110)
E                   	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2059)
E                   	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2048)
E                   	at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:49)
E                   	at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:737)
E                   	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2061)
E                   	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2082)
E                   	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2101)
E                   	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2126)
E                   	at org.apache.spark.rdd.RDD$$anonfun$collect$1.apply(RDD.scala:945)
E                   	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
E                   	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
E                   	at org.apache.spark.rdd.RDD.withScope(RDD.scala:363)
E                   	at org.apache.spark.rdd.RDD.collect(RDD.scala:944)
E                   	at org.apache.spark.sql.execution.SparkPlan.executeCollect(SparkPlan.scala:299)
E                   	at org.apache.spark.sql.Dataset$$anonfun$collectToPython$1.apply(Dataset.scala:3257)
E                   	at org.apache.spark.sql.Dataset$$anonfun$collectToPython$1.apply(Dataset.scala:3254)
E                   	at org.apache.spark.sql.Dataset$$anonfun$53.apply(Dataset.scala:3364)
E                   	at org.apache.spark.sql.execution.SQLExecution$$anonfun$withNewExecutionId$1.apply(SQLExecution.scala:78)
E                   	at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:125)
E                   	at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:73)
E                   	at org.apache.spark.sql.Dataset.withAction(Dataset.scala:3363)
E                   	at org.apache.spark.sql.Dataset.collectToPython(Dataset.scala:3254)
E                   	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
E                   	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
E                   	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
E                   	at java.lang.reflect.Method.invoke(Method.java:498)
E                   	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
E                   	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
E                   	at py4j.Gateway.invoke(Gateway.java:282)
E                   	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
E                   	at py4j.commands.CallCommand.execute(CallCommand.java:79)
E                   	at py4j.GatewayConnection.run(GatewayConnection.java:238)
E                   	at java.lang.Thread.run(Thread.java:748)
E                   Caused by: org.apache.spark.SparkException: Failed to execute user defined function($anonfun$4: (struct<type:tinyint,size:int,indices:array<int>,values:array<double>>) => double)
E                   	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.processNext(Unknown Source)
E                   	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
E                   	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anonfun$13$$anon$1.hasNext(WholeStageCodegenExec.scala:636)
E                   	at org.apache.spark.sql.execution.SparkPlan$$anonfun$2.apply(SparkPlan.scala:255)
E                   	at org.apache.spark.sql.execution.SparkPlan$$anonfun$2.apply(SparkPlan.scala:247)
E                   	at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$24.apply(RDD.scala:836)
E                   	at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$24.apply(RDD.scala:836)
E                   	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
E                   	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)
E                   	at org.apache.spark.rdd.RDD.iterator(RDD.scala:288)
E                   	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
E                   	at org.apache.spark.scheduler.Task.run(Task.scala:121)
E                   	at org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)
E                   	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)
E                   	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)
E                   	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
E                   	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
E                   	... 1 more
E                   Caused by: java.util.NoSuchElementException: Failed to find a default value for threshold
E                   	at org.apache.spark.ml.param.Params$$anonfun$getOrDefault$2.apply(params.scala:780)
E                   	at org.apache.spark.ml.param.Params$$anonfun$getOrDefault$2.apply(params.scala:780)
E                   	at scala.Option.getOrElse(Option.scala:121)
E                   	at org.apache.spark.ml.param.Params$class.getOrDefault(params.scala:779)
E                   	at org.apache.spark.ml.PipelineStage.getOrDefault(Pipeline.scala:42)
E                   	at org.apache.spark.ml.param.Params$class.$(params.scala:786)
E                   	at org.apache.spark.ml.PipelineStage.$(Pipeline.scala:42)
E                   	at org.apache.spark.ml.classification.LogisticRegressionParams$class.getThreshold(LogisticRegression.scala:122)
E                   	at org.apache.spark.ml.classification.LogisticRegressionModel.getThreshold(LogisticRegression.scala:991)
E                   	at org.apache.spark.ml.classification.LogisticRegressionModel.raw2prediction(LogisticRegression.scala:1174)
E                   	at org.apache.spark.ml.classification.ProbabilisticClassificationModel$$anonfun$4.apply(ProbabilisticClassifier.scala:136)
E                   	at org.apache.spark.ml.classification.ProbabilisticClassificationModel$$anonfun$4.apply(ProbabilisticClassifier.scala:136)
E                   	... 18 more

venv/lib/python3.6/site-packages/py4j/protocol.py:328: Py4JJavaError
```

If I understand correctly, deserializing a LogisticRegressionModel with Spark is using [LogisticRegressionOp](https://github.com/combust/mleap/blob/80aefcfa0b6d298c4d703eff8ee9ebf035fc0ef0/mleap-spark/src/main/scala/org/apache/spark/ml/bundle/ops/classification/LogisticRegressionOp.scala#L56) from mleap-spark. It will try to look for `threshold` parameter first, if it is not there, nothing will be set. 

Since scikit-learn LogisticRegressionModel doesn't use `threshold` to make predictions while spark LogisticRegressionModel does, this error will be thrown when we try to run prediction on a spark deserialized LogisticRegressionModel. I saw that [LogisticRegressionOp](https://github.com/combust/mleap/blob/d276c1c800277a1edcf30dda6e8559c7413b21f8/mleap-runtime/src/main/scala/ml/combust/mleap/bundle/ops/classification/LogisticRegressionOp.scala#L51-L53) from mleap-runtime does set a default value for `threshold` if it is not found, can we do the same thing in mleap-spark to set a default value for `threshold`?

### PR DESCRIPTION
**Pull request by ancasarb**
Fixes #614

### CODE REVIEW

**Comment by lucagiovagnoli** [comment]
@tinaxi for quick review too

**Comment by tinaxi** [comment]
@ancasarb @lucagiovagnoli LGTM! Since [mleap-runtime](https://github.com/combust/mleap/blob/d276c1c800277a1edcf30dda6e8559c7413b21f8/mleap-runtime/src/main/scala/ml/combust/mleap/bundle/ops/classification/LogisticRegressionOp.scala#L51-L53) uses the same way to get the threshold, this should fix the deserialization bug that I was seeing.

**Comment by tinaxi** [comment]
> Thanks for taking care of this! Looks great. Ship-it
> 
> Wondering if there are tests for the `store()` and `load()`? To catch these in the future.
> I found these tests but I'm not sure if it's the right spot: https://github.com/combust/mleap/blob/80aefcfa0b6d298c4d703eff8ee9ebf035fc0ef0/mleap-runtime/src/test/scala/ml/combust/mleap/runtime/transformer/classification/LogisticRegressionSpec.scala

@lucagiovagnoli I found there is a parity spec for LogisticRegression model: https://github.com/combust/mleap/blob/d276c1c800277a1edcf30dda6e8559c7413b21f8/mleap-spark/src/test/scala/org/apache/spark/ml/parity/classification/LogisticRegressionParitySpec.scala

But it is creating the logistic regression model in spark side from the beginning, which has the threshold attribute being set. However, the bug I saw only happens when we create logistic regression model using scikit learn (no threshold option) and deserialize it in spark. So this test cannot catch it.

File: mleap-spark/src/main/scala/org/apache/spark/ml/bundle/ops/classification/LogisticRegressionOp.scala [hunk 1]
```diff
@@ -53,9 +53,13 @@ class LogisticRegressionOp extends SimpleSparkOp[LogisticRegressionModel] {
...
           coefficients = Vectors.dense(model.value("coefficients").getTensor[Double].toArray),
           intercept = model.value("intercept").getDouble)
-        model.getValue("threshold").
-          map(t => lr.setThreshold(t.getDouble)).
-          getOrElse(lr)
+
+        // default threshold is 0.5 for both Spark and Scikit-learn
+        val threshold = model.getValue("threshold")
+          .map(value => value.getDouble)
+          .getOrElse(0.5)
```
[the remark this answers is not in the document]
Comment (ancasarb) [reply mleap-spark/src/main/scala/org/apache/spark/ml/bundle/ops/classification/LogisticRegressionOp.scala]: updated with the constant, thank you!

**Comment by ancasarb** [comment]
Yes, we have parity tests like LogisticRegressionParitySpec to check the store and load but as @tinaxi mentioned we don't have parity tests that go from scikit-learn->mleap->spark. There's this ticket to try to do that eventually https://github.com/combust/mleap/issues/316.

**Pull request merged**